### DISCUSSION TITLE
Creating my own `U8` type

### DISCUSSION THREAD

**Discussion by jymchng** [modular/modular#715] [Q&A]
Following the tutorial from the [docs](https://docs.modular.com/mojo/notebooks/BoolMLIR.html#adding-functionality-with-mlir) on creating a new boolean using MLIR, I tried to create my own new `U8` type.

Yet, it is difficult to do so because the constructor (`__init__`) will need to take `Int` as a parameter and I'm not too sure how to check for the value passed into the constructor such that it doesn't overflow etc.

```
struct U8:
   var value: __mlir_type.ui8

   fn __init__(inout self, value: Int):
      self.value = value


fn main():
   let a = U8(__mlir_attr.`55`)
```

In the MLIR [docs](https://mlir.llvm.org/docs/Dialects/IndexOps/#indexcmp-indexcmpop) for `index.cmp` in the `index` dialect, the `cmp` operation is as follows:
```
operation ::= `index.cmp` `` $pred `(` $lhs `,` $rhs `)` attr-dict
```

In the Mojo docs, the 'conversion' to Mojo's syntax is as follows:
```
fn __eq__(self, rhs: OurBool) -> Self:
    let lhsIndex = __mlir_op.`index.casts`[_type : __mlir_type.index](
        self.value
    )
    let rhsIndex = __mlir_op.`index.casts`[_type : __mlir_type.index](
        rhs.value
    )
    return Self(
        __mlir_op.`index.cmp`[
            pred : __mlir_attr.`#index<cmp_predicate eq>`
        ](lhsIndex, rhsIndex)
    )
```

How does

```
`index.cmp` `` $pred `(` $lhs `,` $rhs `)` attr-dict`
```

get converted to

```
__mlir_attr.`#index<cmp_predicate eq>`
```

**Comment by jymchng** [comment]
How is `Int` represented in MLIR?

**Comment by alfie-nsugh** [comment]
I think it has to do with this:
https://mlir.llvm.org/docs/LangRef/#dialect-attribute-values
